Tweak URI-references and be smart with Strings

duglin opened the pull request:
I can't think of a good reason why URI-references should be allowed to
be blank since that seems about as useful as the property not being
there at all.

Closes #473

Signed-off-by: Doug Davis <dug@us.ibm.com>

duglin commented:
I think this help address some of https://github.com/cloudevents/spec/pull/456  too

cneijenhuis commented on primer.md:
```diff
@@ -235,6 +236,54 @@ general, this is true for all CloudEvents attributes as well.
+Attributes of type URI-reference have quite a large set of valid values that
+can be used. For example, it is possible that the `source` attribute could be
+as short as `home` or could be more meaningful, such as
+`com.example.my-component`. The CloudEvents specification allows for a wide
```
`example.com/my-component` sounds more like a proper URI-ref to me (unless `my-component` is a new TLD I haven't heard of yet 😉 )

duglin replied:
fixed :-)

cneijenhuis commented on primer.md:
```diff
@@ -235,6 +236,54 @@ general, this is true for all CloudEvents attributes as well.
+interoperability while still being flexible enough to be used in unique
+situations. So, CloudEvent producers are encouraged to only use values that
+have the best chance of being properly processed by all possible formats
+and transports that their CloudEvents might be used with.
```
For SaaS publishers, their CloudEvents may end up being used with any format or transport. It'd be good to know the minimum one can rely on. ASCII? URL encoding?

duglin replied:
I thought about that too, but then I noticed that define Strings as `Sequence of printable Unicode characters`
which to me removed a fair about of the reasons for needing to URL encode stuff. Plus I didn't think we'd want to get into when, or when not, to encode Strings since then we'd need a mechanism by which consumers would have to know when they're encoded.  That's when I decided "just don't be dumb" was good enough.

But if someone has any ideas...

rperelma commented on primer.md:
```diff
@@ -235,6 +236,54 @@ general, this is true for all CloudEvents attributes as well.
+as short as `home` or could be more meaningful, such as
+`http://home.example.com`. The CloudEvents specification allows for a wide
+range of values to allow for it to be used in as any use cases as possible,
+without undo burden on implementations.
```
s/undo/undue

duglin replied:
LOL fixed

deissnerk commented on primer.md:
```diff
@@ -235,6 +236,54 @@ general, this is true for all CloudEvents attributes as well.
+themselves based on their needs and expected usage of their events.
+
+Additionally, the CloudEvents specification does not define a `baseURI`
+to be used along with the URI-reference type attributes. This is because,
```
I wonder, if it prohibits that event producers would define a base URI in their documentation?

duglin replied:
I don't think we can stop anyone from doing so... but I'd question the need for it  :-)

deissnerk commented on primer.md:
```diff
@@ -235,6 +236,54 @@ general, this is true for all CloudEvents attributes as well.
+unless otherwise stated, the URI-references are not meant to be normalized
+or used for any other purpose other than comparing its string value against
+some other URI-reference string value for equality. The use of URI-reference,
+instead of just a String, was done to encourage the use of meaningful values
```
Referring to a complex specification such as [RFC3986](https://tools.ietf.org/html/rfc3986) just to encourage people to chose meaningful values while ignoring most of the specification, seems a bit strange.

duglin replied:
maybe but I think there is something to be said for encouraging some level of consistency, but still allowing a lot of freedom.  Kind of like a best-practice.  But, having said that, this is just my memory of our previous chats and if people think I'm remembering incorrectly, or they just don't want to even mention this aspect of it, I'd be ok with removing this sentence if that would help

duglin commented:
up updated based on today's call.  Let me know what you think

deissnerk commented:
Thank you @duglin! LGTM

evankanderson commented on spec.md:
```diff
@@ -363,6 +368,7 @@ on the definition of OPTIONAL.
   for more information.
 - Constraints:
   - OPTIONAL
+  - If present, MUST be a non-empty URI-reference
```
I'm more skeptical of the value of a URI-reference here; I sort of understand the thinking that we want to allow a subset of general strings as `source` attributes even if they don't conform to the
```URI           = scheme ":" hier-part [ "?" query ] [ "#" fragment ]```
production. Since `schemaurl` is OPTIONAL rather than required, why not require that the schema URL be an actual URI and not relative to some (unknown) base?

deissnerk replied:
@evankanderson I think that's part of #475.

duglin commented:
moved the "RECOMMENDED" line into the constraints section and deleted the rest of the paragraph - per today's call.

Approved with the above change on the 8/15 call

Can I get an LGTM to make sure I didn't mess anything up before I merge?

cneijenhuis commented on spec.md:
```diff
@@ -251,6 +251,8 @@ The following attributes are REQUIRED to be present in all CloudEvents:
 - Constraints:
   - REQUIRED
+  - MUST be a non-empty URI-reference
+  - RECOMMENDED that an absolute URI be used
```
Disclaimer: I'm not a native English speaker. `be used` sounds strange to me here.
Either try `is used` or maybe `An absolute URI is RECOMMENDED`?

duglin replied:
works for me!

duglin commented:
thanks for the review @cneijenhuis 

merging.  Approved on the 8/15 call
